File: 70-climbing-stairs/climbing-stairs.py
class Solution:
    def solve(self, n:int, memo: dict[int,int]) -> int:
        if n == 0 or n == 1:
            return 1
        
        if n not in memo:
            memo[n] = self.solve(n-1, memo) + self.solve(n-2, memo)
        return memo[n]
    def climbStairs(self, n: int) -> int:
        memo = {}
        return self.solve(n, memo)

File: 70-climbing-stairs/climbing-stairs.cpp
class Solution {
public:
    int solve(int n, vector<int> &memo){
        if (n == 0 || n == 1) return 1;
        if (n < 0) return 0;
        if (memo[n] != -1) return memo[n];

        return memo[n] = solve(n-1, memo) + solve(n-2, memo);
    }
    int climbStairs(int n) {
        vector<int> dp(n+1, -1);
        return solve(n, dp);
    }
};